Publish docs to eui.elastic.co

JoshMock opened the pull request:
### Summary

Adds scripts and a Jenkins job definition that build and deploy EUI docs to an internal, Bekitzur-managed GCS bucket, soon to replace the work of publishing docs to Github Pages.

See https://eui.elastic.co for an example of the docs published for the most current version of EUI.

### Checklist

N/A; I didn't change any EUI code to do this. 😎 

- [ ] ~Check against **all themes** for compatibility in both light and dark modes~
- [ ] ~Checked in **mobile**~
- [ ] ~Checked in **IE11** and **Firefox**~
- [ ] ~Props have proper **autodocs**~
- [ ] ~Added **documentation** examples~
- [ ] ~Added or updated **jest tests**~
- [ ] ~Checked for **breaking changes** and labeled appropriately~
- [ ] ~Checked for **accessibility** including keyboard-only and screenreader modes~
- [ ] ~A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately~

nyurik commented on scripts/deploy/deploy_docs:
```diff
@@ -0,0 +1,103 @@
+    gsutil cp \
+      -a public-read \
+      -z html \
+      -h "Cache-Control:public, max-age=3600" \
```
we could reduce this to 5 min, but no biggie

nyurik commented on scripts/deploy/redirect_template.html:
```diff
@@ -0,0 +1,8 @@
+<!DOCTYPE html>
+<html>
```
after some pondering, i think javascript method is better here - no 2 second delay (annoying). See https://en.wikipedia.org/wiki/Meta_refresh#For_redirection -- use `<body onload="window.location = 'https://example.com/'">Redirecting to https://example.com/</body>`

JoshMock replied:
Yeah, I was wondering about that, too. It wouldn't work if JS was disabled, but we might be able to keep the meta tag there as a backup in that case. I'll make this change.

nyurik replied:
true, plus if JS is disabled, who cares - EUI wouldn't work anyway :)

snide commented on scripts/deploy/deploy_docs:
```diff
@@ -0,0 +1,103 @@
+  # if this is the highest release number to date, set redirect from root to
+  # this version's directory
+  if [[ "v$CURRENT_RELEASE" == "$GIT_BRANCH" ]]; then
+    echo "This is the current release. Redirecting https://eui.elastic.co/ to path https://eui.elastic.co/$GIT_BRANCH/"
```
Really excellent point. I think we'd want to do a couple things.

1. Put a no crawl / robots.txt bit on anything that is not the latest. We don't want people hitting old versions unless they are specifically asking for it.
2.  I would default the latest released version and to not include the version number in the URL (it would be fine if the lastest lived in two places, / and /vX.X.X/). To your point, we would prefer old links to hit new versions of code.

snide replied:
For reference, I think this really hurts the actual elastic.co docs when you consider that most people are coming to them from outside places (google, slack..etc)

thompsongl replied:
Instead of a redirect, what if we had the latest tag deployment copied to a latest/production bucket? [ems-landing-page](https://github.com/elastic/ems-landing-page/blob/master/deployProduction.sh) does something similar with their staging & prod environments

JoshMock replied:
So, to clarify: when, say, `v19.0.0` comes out we'd want to publish it to `eui.elastic.co/v19.0.0/` but also overwrite `eui.elastic.co/` to serve the same files? And then, I assume, put a [canonical meta tag](https://blog.alexa.com/canonical-url/) redirecting one to the other?

If so, I'd be responsible for publishing the latest version to two places instead of just redirecting to latest, which is what it does now. And you all would be responsible for maintaining any robots.txt rules in the repo's `/docs` directory.

thompsongl replied:
Yes, that sounds correct to me

snide replied:
Yep. Sounds good!

JoshMock replied:
Great. I meant to say, too, that the canonical meta tag would be your responsibility as well, since it's part of the actual docs markup.

I'll update this script to publish latest docs to both locations.

JoshMock commented on scripts/deploy/build_docs:
```diff
@@ -0,0 +1,23 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
```
I copied this from other Bekitzur scripts in another public repo, which I assume @nyurik put there when he wrote them. I'm honestly not too familiar with our policies about when/where to post licenses.

chandlerprall replied:
Good catch @thompsongl!

We don't have the Elastic License text in EUI (nor would we want to); we will be adding the Apache 2.0 license header to all files soon (likely February), including here. Might as well put it in now.

```
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
```

JoshMock commented:
Updated the script to publish latest docs to both locations, and fixed the software license comments! Ready for one final look, please. :pray:

One thing of note: once this merges, we'll still need to deploy old version tags of the docs manually. These scripts only live in master, so it'll only work on upcoming tags. Let me know if you'd like older versions to get published and we can work out how to go about it.

nyurik commented:
Looks good, but please double check the role under which this job will run, e.g. in the landing page see https://github.com/elastic/ems-landing-page/blob/master/.ci/jobs/defaults.yml#L38 for the example.

JoshMock commented:
Good catch @nyurik. Can you give one last approval?

nyurik commented:
@JoshMock looks good, thx!  The only possible issue might be if you want vault role to be specific to the bucket publishing job only, you may want to put it into the job itself, rather than default, but this is not a biggie.

snide commented:
> One thing of note: once this merges, we'll still need to deploy old version tags of the docs manually. These scripts only live in master, so it'll only work on upcoming tags. Let me know if you'd like older versions to get published and we can work out how to go about it.

This maps against minor releases for Kibana since 6.8 (which is our last support). If we could do these for the past it would be fantastic. One thing we'll need to do on our side is figure out a way to surface this in the docs as a switcher, but that's on us.

* 18.2.1
* 14.5.1
* 13.6.1
* 12.2.1
* 10.4.2
* 6.10.9

nyurik commented:
If we could craft a POST request to Jenkins to trigger a job with a specific parameter (tag version), once everything is working we could write a very simple script that just makes a POST request for each tag that exists, possibly waiting for 30seconds between each call so not to overload the system). This should backfill everything without much effort.

JoshMock commented:
Would that work knowing that the job and the script it runs aren't available outside the `master` branch?

nyurik commented:
hmm... well, the job itself will work just fine because JJBB will only import jobs from master into Jenkins. The issue is that the job then calls a script in the current branch of the GIT, and that script won't exist outside of master.  I can see two ways around it:
* as part of bootstrapping, switch to the master branch, and pass needed git branch as a parameter so that the script, once started, could switch the branch back, or as an alternative, copy needed scripts from the master branch to some temp dir without switching the current branch.
* Move the script into the job itself (job already contains a part of the bootstraping script, but it could in theory contain more).  I think this approach is worse the the one above.

JoshMock commented:
Both approaches are tricky because the script calls itself, indirectly, by re-invoking from within a Docker container.

For the first solution, you wouldn't be able to switch branches after starting the script because then the Docker re-invocation would fail.

And for the second solution, you'd have to split the work up into two jobs, with the second job depending on build artifacts that are generated in the first job, which we don't typically do.

JoshMock commented:
For those reasons, manually copying the scripts and then switching branches is the easiest approach. Still not optimal, of course, but it's a one-time operation that wouldn't be hard to automate with a throwaway script if we needed to do it for all release tags.
